src/mixins/themeable.ts

```diff
@@ -0,0 +1,154 @@
+
+/**
+ * Private map for the widgets themeClasses.
+ */
+const themeClassesMap = new WeakMap<ThemeableMixin<{}>, AppliedClasses<any>>();
+
+function addClassNameToCSSModuleClassNames(cssModuleClassNames: CSSModuleClassNames, classList: StringIndexedObject, className: string) {
+	if (classList.hasOwnProperty(className)) {
+		// split out the classname because css-module composition combines class names with a space
+		const generatedClassNames: string[] = classList[className].split(' ');
+		generatedClassNames.forEach((generatedClassName) => {
+			cssModuleClassNames[generatedClassName] = true;
+		});
+	}
+}
+
+function negatePreviousClasses<T>(previousClasses: AppliedClasses<T>, newClasses: AppliedClasses<T>) {
+	return Object.keys(previousClasses).reduce((newAppliedClasses, className: keyof T) => {
+		const oldCSSModuleClassNames: CSSModuleClassNames = previousClasses[className] || {};
+
+		const negatedCSSModuleClassNames = Object.keys(oldCSSModuleClassNames).reduce((newCSSModuleClassNames, oldCSSModuleClassName) => {
+			const currentClassNameFlag = oldCSSModuleClassNames[<keyof T> oldCSSModuleClassName];
+			// If it's true it needs to be negated and passed along, If it's false,
+			// don't return it as maquette will already have removed it.
+			if (currentClassNameFlag) {
+				newCSSModuleClassNames[oldCSSModuleClassName] = false;
+			}
+			return newCSSModuleClassNames;
+		}, <CSSModuleClassNames> {});
+
+		const calculatedClassNameMap = assign({}, negatedCSSModuleClassNames, newClasses[className]);
+		newAppliedClasses[className] = calculatedClassNameMap;
+
+		return newAppliedClasses;
+	}, <AppliedClasses<T>> {});
+}
+
+function generateThemeClasses<I, T>(instance: Themeable<I>, baseTheme: T, theme: {} = {}, overrideClasses: {} = {}) {
+	return Object.keys(baseTheme).reduce((newAppliedClasses, className: keyof T) => {
+		const newCSSModuleClassNames: CSSModuleClassNames = {};
+
+		let themeClassSource: {} = baseTheme;
+		if (theme.hasOwnProperty(className)) {
+			themeClassSource = theme;
+		}
+
+		addClassNameToCSSModuleClassNames(newCSSModuleClassNames, themeClassSource, className);
+		overrideClasses && addClassNameToCSSModuleClassNames(newCSSModuleClassNames, overrideClasses, className);
+		newAppliedClasses[className] = newCSSModuleClassNames;
+
+		return newAppliedClasses;
+	}, <AppliedClasses<T>> {});
+}
+
+function updateThemeClassesMap<I, T>(instance: Themeable<I>, newThemeClasses: AppliedClasses<T>) {
+	if (themeClassesMap.has(instance)) {
+		const previousThemeClasses = themeClassesMap.get(instance);
+		themeClassesMap.set(instance, negatePreviousClasses(previousThemeClasses, newThemeClasses));
+	} else {
+		themeClassesMap.set(instance, newThemeClasses);
+	}
+}
+
+function onPropertiesChanged<I>(instance: Themeable<I>, { theme, overrideClasses }: ThemeableProperties, changedPropertyKeys: string[]) {
+	const { theme: propTheme, overrideClasses: propOverrideClasses } = instance.properties;
```

Comment:
Good spot, but I don't feel it really matters either way. Can change if you're especially passionate about it.